dbstructure/query.sql

```diff
@@ -0,0 +1,483 @@
+       with
+       views as (
+         select
+           n.nspname   as view_schema,
+           c.relname   as view_name,
+           r.ev_action as view_definition
+         from pg_class c
+         join pg_namespace n on n.oid = c.relnamespace
+         join pg_rewrite r on r.ev_class = c.oid
+         where c.relkind in ('v', 'm') and n.nspname = ANY ($1)
+       ),
+       removed_subselects as(
+         select
+           view_schema, view_name,
+           regexp_replace(view_definition,
+             -- "result" appears when the subselect is used inside "case when", see
+             -- `authors_have_book_in_decade` fixture
+             -- "resno"  appears in every other case
+             case when (select pgv_num from pg_version) < 100000
+                then ':subselect {.*?:constraintDeps <>} :location \d+} :res(no|ult)'
+                else ':subselect {.*?:stmt_len 0} :location \d+} :res(no|ult)'
+             end,
+           '', 'g') as x
+         from views
+       ),
+       target_lists as(
+         select
+           view_schema, view_name,
+           regexp_split_to_array(x, 'targetList') as x
+         from removed_subselects
+       ),
+       last_target_list_wo_tail as(
+         select
+           view_schema, view_name,
+           (regexp_split_to_array(x[array_upper(x, 1)], ':onConflict'))[1] as x
+         from target_lists
+       ),
+       target_entries as(
+         select
+           view_schema, view_name,
+           unnest(regexp_split_to_array(x, 'TARGETENTRY')) as entry
+         from last_target_list_wo_tail
+       ),
+       results as(
+         select
+           view_schema, view_name,
+           substring(entry from ':resname (.*?) :') as view_colum_name,
+           substring(entry from ':resorigtbl (.*?) :') as resorigtbl,
+           substring(entry from ':resorigcol (.*?) :') as resorigcol
+         from target_entries
+       )
+       select
+         sch.nspname as table_schema,
+         tbl.relname as table_name,
+         col.attname as table_column_name,
+         res.view_schema,
+         res.view_name,
+         res.view_colum_name,
+         source_column.c1 as src_source,
+         view_column.c2 as src_view
+       from results res
+       join pg_class tbl on tbl.oid::text = res.resorigtbl
+       join pg_attribute col on col.attrelid = tbl.oid and col.attnum::text = res.resorigcol
+       join pg_namespace sch on sch.oid = tbl.relnamespace,
+       lateral (
+            select c1
+            from columns c1
+            where
+                sch.nspname = c1.col_schema
+                and tbl.relname = c1.col_table_name
+                and col.attname = c1.col_name
+       ) source_column,
+       lateral (
+            select c2
+            from columns c2
+            where
+                res.view_schema = c2.col_schema
+                and res.view_name = c2.col_table_name
+                and res.view_colum_name = c2.col_name
+       ) as view_column
```

Comment:
These laterals are what's making the query costly. I think we would get the same time(or better) as the old query if we remove them. Do you think it could be possible to go back to making the matching of the columns and source_columns in Haskell code?

Reply:
The number of columns can get big. I don't see a good way for getting them in the query for now.

Reply:
(In case you'd like to better test this. I can send you a sample big schema to your email.)

Reply:
@steve-chavez Very interesting, thank you for investigating!! Do you think it's an issue with the amount of data that we need to transfer with this approach?

I think there is a lot of potential for optimizing the big query, would be exciting to see if we can push through and get this much faster - which should be possible in principle I believe. If there is a larger data set that you can share, that would be very helpful!